sklearn/utils/validation.py

```diff
@@ -80,7 +80,15 @@ def as_float_array(X, copy=True, force_all_finite=True):
     elif X.dtype in [np.float32, np.float64]:  # is numpy array
         return X.copy('F' if X.flags['F_CONTIGUOUS'] else 'C') if copy else X
     else:
-        return X.astype(np.float32 if X.dtype == np.int32 else np.float64)
+        return X.astype(np.float32 if X.dtype in (
+                                                  np.int8,
```

Comment:
If you do this:
```python
a = np.int(0)
a.dtype.itemsize
```
then you get an error
However, doing `X.dtype.kind in 'uib' and X.dtype.itemsize <= 4` works, so I changed that in the latest commit. If that commit is ok, I'll clean up the history to only have 2 commits.